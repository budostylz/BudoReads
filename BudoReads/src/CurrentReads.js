/**
  * TODO: CurrentReads Component
*/


import React from 'react'
import PropTypes from 'prop-types'

const CurrentReads = (props) => (<div className="bookshelf">
    <h2 className="bookshelf-title">Currently Reading</h2>
    <div className="bookshelf-books">
        <ol className="books-grid">
            {/*console.log('CurrentReads props', props.currentBooks.filter(book => book.shelf === 'currentlyReading'))*/}
            {props.currentBooks.filter(book => book.shelf === 'currentlyReading').map(book => (<li key={book.id}>
                <div className="book" id={book.id}>
                    <div className="book-top">
                        <div className="book-cover"
                            style={{ width: 128, height: 188, backgroundImage: `url("${(book.imageLinks) ? book.imageLinks.thumbnail : ''}")` }}>
                        </div>
                        <div className="book-shelf-changer">
                            <select
                                defaultValue='currentlyReading'
                                onChange={props.selectBookShelf}

                            >
                                <option value="move" disabled>Move to...</option>
                                <option value="currentlyReading">Currently Reading</option>
                                <option value="wantToRead">Want to Read</option>
                                <option value="read">Read</option>
                                <option value="none">None</option>
                            </select>
                        </div>
                    </div>
                    <div className="book-title">{book.title}</div>
                    <div className="book-authors">
                        {(book.authors && book.authors.length > 0) ? book.authors.toString() : 'No Author(s)'}
                    </div>
                </div>
            </li>))}


        </ol>
    </div>
</div>)




CurrentReads.propTypes = {
    currentBooks: PropTypes.array.isRequired,
    selectOption: PropTypes.string.isRequired,
    selectBookShelf: PropTypes.func.isRequired

}



export default CurrentReads