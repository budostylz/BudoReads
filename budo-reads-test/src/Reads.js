import React from 'react'


const Reads = (props) =>
    (<div className="bookshelf">
        <h2 className="bookshelf-title">Read</h2>
        <div className="bookshelf-books">
            <ol className="books-grid">
                {/*console.log('Reads props', props)*/}
                {props.readBooks.filter(book => book.shelf === 'read').map(book => (<li key={book.id}>
                    <div className="book" id={book.id}>
                        <div className="book-top">
                            <div className="book-cover"
                                style={{ width: 128, height: 188, backgroundImage: `url("${(book.imageLinks) ? book.imageLinks.thumbnail : ''}")` }}>
                            </div>
                            <div className="book-shelf-changer">
                                <select
                                    defaultValue='read'
                                    onChange={props.selectBookShelf}
                                >
                                    <option value="move" disabled>Move to...</option>
                                    <option value="currentlyReading">Currently Reading</option>
                                    <option value="wantToRead">Want to Read</option>
                                    <option value="read" defaultValue={'selected'}>Read</option>
                                    <option value="none">None</option>
                                </select>
                            </div>
                        </div>
                        <div className="book-title">{book.title}</div>
                        <div className="book-authors">
                            {(book.authors && book.authors.length > 0) ? book.authors.toString() : 'No Author(s)'}
                        </div>
                    </div>
                </li>))}

            </ol>
        </div>
    </div>)


export default Reads