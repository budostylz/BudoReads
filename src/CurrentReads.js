/**
  * TODO: CurrentReads Component
*/


import React from 'react'
import Select from 'react-select'

const CurrentReads = (props) => (<div className="bookshelf">
    <h2 className="bookshelf-title">Currently Reading</h2>
    <div className="bookshelf-books">
        <ol className="books-grid">
            {console.log('CurrentReads props', props)}
            {props.currentBooks.map(book => (<li key={book.id}>
                <div className="book">
                    <div className="book-top">
                        <div className="book-cover" style={{ width: 128, height: 188, backgroundImage: `url("${book.imageLinks.thumbnail}")` }}></div>
                        <div className="book-shelf-changer">
                            <select>
                                <option value="move" disabled>Move to...</option>
                                <option value="currentlyReading">Currently Reading</option>
                                <option value="wantToRead">Want to Read</option>
                                <option value="read" defaultValue={'selected'}>Read</option>
                                <option value="none">None</option>
                            </select>
                        </div>
                    </div>
                    <div className="book-title">{book.title}</div>
                    <div className="book-authors">
                        {(book.authors.length > 0) ? book.authors.toString() : 'No Author(s)'}
                    </div>
                </div>
            </li>))}

        </ol>
    </div>
</div>)










export default CurrentReads