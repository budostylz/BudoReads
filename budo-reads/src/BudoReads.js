/**
  * TODO: BudoReads Component: Parent component that houses state and sends props to child components.
*/

import React, { Component } from 'react'
import { Route } from 'react-router-dom'
import CurrentReads from './CurrentReads.js'
import WantToReads from './WantToReads.js'
import Reads from './Reads.js'
import Nav from './Nav.js'
import Search from './Search.js'
import * as BooksAPI from './BooksAPI'


class BudoReads extends Component {


    /**
      * TODO: 3 pieces of state:
      * myBooks: Collection of books that are pulled from API and updated by user via user interface.
      * searchResults: Collection of books that are pulled from search API.
      * selectOption: Tracks user shelf selection.
    */

    state = {

        myBooks: [],
        searchResults: [],
        selectOption: ''


    }


    /**
      * TODO: componentDidMount()
      * Pulls books from API results using getAll() after DOM is rendered in lifecycle.
      * Sets myBooks[] within state
      * 
    */

    componentDidMount() {//getAll
        BooksAPI.getAll()
            .then((books) => {
                this.setState(currentState => ({
                    myBooks: books
                }))
            })
    }

    /**
         * TODO: updateBookShelf()
         * Updates book shelf base on user selection.
         * Retrieves user selection 'currently reading', 'want to read' and 'read'.
         * Updates API base on user selection.
         * Updates myBooks and selectOption within state base on user selection.
       */

    updateBookShelf = (e) => {//update
        const { myBooks } = this.state;
        const shelf = e.target.options[e.target.options.selectedIndex].value;
        const bookID = e.target.parentElement.parentElement.parentElement.getAttribute('id');
        const book = myBooks.filter(book => book.id === bookID)[0]; //show me a better way to not use an index[0] if possible
        const newBookSet = myBooks.filter(book => book.id !== bookID);

        //update API
        BooksAPI.update(book, shelf)
            .then((book) => {
                BooksAPI.get(bookID)
                    .then((book) => {
                        newBookSet.push(book)
                        this.setState(currentState => ({
                            myBooks: newBookSet,
                            selectOption: shelf

                        }))
                    })
            })


    }

    /**
        * TODO: searchBooks()
        * Search books by pulling from search API base on user query.
        * Handles search errors.
        * Update state if proper search result is pulled.
      */


    searchBooks = (e) => {//search
        const query = e.target.value;
        if (query.trim().length > 0) {
            BooksAPI.search(query)
                .then((result) => {
                    if (result) {
                        if (result.error) {
                            this.setState(currentState => ({
                                searchResults: []
                            }))
                        } else {
                            this.setState(currentState => ({
                                searchResults: result
                            }))
                        }
                    } else {
                        this.setState(currentState => ({
                            searchResults: []
                        }))

                    }
                })

        } else {
            this.setState(currentState => ({
                searchResults: []

            }))
        }



    }

    /**
       * TODO: addBookToShelf()
       * Add books to shelf from search page
       * Retrieves book from API
       * Updates book shelf in API
       * Updates book shelf in state
       * 
     */
    addBookToShelf = (e) => {
        const shelf = e.target.options[e.target.options.selectedIndex].value;
        if (shelf !== 'move') {

            const { myBooks } = this.state;
            const bookID = e.target.parentElement.parentElement.parentElement.getAttribute('id');
            const newBookSet = myBooks.filter(book => book.id !== bookID);

            //console.log(shelf, bookID, newBookSet)

            BooksAPI.get(bookID)
                .then((book) => {
                    BooksAPI.update(book, shelf)
                        .then((updatedBook) => {

                            BooksAPI.get(bookID)
                                .then((updatedBook) => {
                                    //console.log('updated book', updatedBook)

                                    newBookSet.push(updatedBook)

                                    this.setState(currentState => ({
                                        myBooks: newBookSet,
                                        selectOption: shelf

                                    }))

                                })
                        })
                })

        }
    }


    /**
       * TODO: clearSearchResults()
       * Clear search results when user navigates back to home page.
       * 
     */

    clearSearchResults = () => {
        this.setState(currentState => ({
            searchResults: []
        }))
    }

    /**
       * TODO: render()
       * Pass props to child components. 
       * Render child components.
       * 
     */

    render() {
        const { myBooks, selectOption, searchResults } = this.state;
        return (<div>
            <Nav
                searchBooks={this.searchBooks}
                clearSearchResults={this.clearSearchResults}

            />
            <Route
                exact path='/'
                render={() => (
                    <div className="list-books">
                        <div className="list-books-content">
                            <div>

                                <CurrentReads
                                    currentBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />
                                <WantToReads
                                    wantToReadBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />
                                <Reads
                                    readBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />


                            </div>
                        </div>

                    </div>)}
            />

            <Route
                path='/search'
                render={({ history }) => (
                    <Search
                        searchBooks={this.searchBooks}
                        searchResults={searchResults}
                        selectOption={selectOption}
                        selectBookShelf={this.addBookToShelf}
                        currentBooks={myBooks}


                    />
                )}
            />


        </div>)
    }
}

export default BudoReads
