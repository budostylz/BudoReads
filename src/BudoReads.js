/**
  * TODO: BudoReads Component
*/

import React, { Component } from 'react'
import { Route } from 'react-router-dom'
import CurrentReads from './CurrentReads.js'
import WantToReads from './WantToReads.js'
import Reads from './Reads.js'
import Nav from './Nav.js'
import Search from './Search.js'
import * as BooksAPI from './BooksAPI'



class BudoReads extends Component {

    state = {

        myBooks: [],
        searchResults: [],
        selectOption: ''


    }

    componentDidMount() {//getAll

        //console.log('componentDidMount')
        BooksAPI.getAll()
            .then((books) => {
                //console.log('books', books)

                this.getBooks(books);

            })
    }



    getBooks = (books) => {

        this.setState(currentState => ({
            myBooks: books

        }))

    }


    updateBookShelf = (e) => {//update

        const { myBooks, selectOption } = this.state;
        const shelf = e.target.options[e.target.options.selectedIndex].value;


        if (selectOption !== shelf) {

            const bookID = e.target.parentElement.parentElement.parentElement.getAttribute('id');
            const book = myBooks.filter(book => book.id === bookID)[0];
            const newBookSet = myBooks.filter(book => book.id !== bookID);

            //console.log('state', this.state)
            //console.log('shelf', shelf)
            //console.log('book id', bookID)
            //console.log('selectedBook', book)
            //console.log('new book set filtered', newBookSet)



            //update API
            BooksAPI.update(book, shelf)
                .then((book) => {
                    //console.log('update book', book)
                    BooksAPI.get(bookID)
                        .then((book) => {
                            //console.log('get updated book', book)

                            newBookSet.push(book)

                            //console.log('new myBooks state', newBookSet)

                            this.setState(currentState => ({
                                myBooks: newBookSet,
                                selectOption: shelf

                            }))
                        })

                })


        }





    }

    removeBookFromCollection = () => {//get

    }


    searchBooks = () => {//search

    }






    render() {

        console.log('state', this.state)

        const { myBooks, selectOption } = this.state;
        return (<div>

            <Route
                exact path='/'
                render={() => (
                    <div className="list-books">
                        <div className="list-books-content">
                            <div>

                                <CurrentReads
                                    currentBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />
                                <WantToReads
                                    wantToReadBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />
                                <Reads
                                    readBooks={myBooks}
                                    selectOption={selectOption}
                                    selectBookShelf={this.updateBookShelf}
                                />

                                <Nav />
                            </div>
                        </div>

                    </div>)}
            />

            <Route
                path='/search'
                render={({ history }) => (
                    <Search />
                )}
            />



        </div>)
    }
}

export default BudoReads