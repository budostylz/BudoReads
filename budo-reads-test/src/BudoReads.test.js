// Link.react.test.js
import React from 'react';
import BudoReads from './BudoReads.js'
import CurrentReads from './CurrentReads.js'
import WantToReads from './WantToReads.js'
import Reads from './Reads.js'
import renderer from 'react-test-renderer';

test('Display Currently Reading, Want to Read and Read Sections', () => {


    const currentlyReadsComponent = renderer.create(
        <CurrentReads
            currentBooks={[
                {
                    id: "8tNBKmPO5UC",
                    title: "Android Fully Loaded",
                    imageLinks: {
                        thumbnail: "http://books.google.com/books/content?id=H8tNBKmPO5UC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
                    },
                    authors: ["Rob Huddleston"]
                }


            ]}
        />);


    const wantToReadsComponent = renderer.create(
        <WantToReads
            wantToReadBooks={[]}
        />);


    const readsComponent = renderer.create(
        <Reads
            readBooks={[]}
        />);


    let tree1 = currentlyReadsComponent.toJSON();
    expect(tree1).toMatchSnapshot();









});