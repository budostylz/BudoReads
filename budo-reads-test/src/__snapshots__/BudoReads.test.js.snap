// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Currently Reading, Want to Read and Read Sections 1`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  >
    Currently Reading
  </h2>
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;

exports[`Display Currently Reading, Want to Read and Read Sections 2`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  >
    Currently Reading
  </h2>
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;

exports[`Display Currently Reading, Want to Read and Read Sections 3`] = `
<div
  className="bookshelf"
>
  <h2
    className="bookshelf-title"
  >
    Currently Reading
  </h2>
  <div
    className="bookshelf-books"
  >
    <ol
      className="books-grid"
    />
  </div>
</div>
`;

exports[`Link changes the class when hovered 1`] = `null`;
